File: Provenance/Controller/iCade/PViCadeMocuteController.m
//
//  PViCadeMocuteController.m
//  Provenance
//
//  Created by Edgar Neto on 8/12/17.
//  Copyright © 2017 James Addyman. All rights reserved.
//

#import "PViCadeMocuteController.h"
#import "PViCadeGamepad.h"
#import "PViCadeGamepadButtonInput.h"
#import "PViCadeGamepadDirectionPad.h"

@implementation PViCadeMocuteController

-(instancetype) init {
    if (self = [super init]) {
        __unsafe_unretained PViCadeController* weakSelf = self;
        self.reader.buttonDown = ^(iCadeState button) {
            switch (button) {
                case iCadeButtonA:
                    [[weakSelf.iCadeGamepad buttonX] buttonPressed];
                    break;
                case iCadeButtonB:
                    [[weakSelf.iCadeGamepad buttonA] buttonPressed];
                    break;
                case iCadeButtonC:
                    [[weakSelf.iCadeGamepad buttonB] buttonPressed];
                    break;
                case iCadeButtonD:
                    [[weakSelf.iCadeGamepad buttonY] buttonPressed];
                    break;
                case iCadeButtonE:
                    [[weakSelf.iCadeGamepad leftShoulder] buttonPressed];
                    break;
                case iCadeButtonF:
                    [[weakSelf.iCadeGamepad rightShoulder] buttonPressed];
                    break;
                case iCadeButtonG:
                    [[weakSelf.iCadeGamepad leftTrigger] buttonPressed];
                    break;
                case iCadeButtonH:
                    [[weakSelf.iCadeGamepad rightTrigger] buttonPressed];
                    break;
                case iCadeJoystickDown:
                case iCadeJoystickLeft:
                case iCadeJoystickRight:
                case iCadeJoystickUp:
                    [[weakSelf.iCadeGamepad dpad] padChanged];
                    break;
                default:
					WLOG(@"Unknown iCade key %0x", button);
                    break;
            }
            if (weakSelf.controllerPressedAnyKey) {
                weakSelf.controllerPressedAnyKey(weakSelf);
            }
        };
        
        self.reader.buttonUp = ^(iCadeState button) {
            switch (button) {
                case iCadeButtonA:
                    [[weakSelf.iCadeGamepad buttonX] buttonReleased];
                    break;
                case iCadeButtonB:
                    [[weakSelf.iCadeGamepad buttonA] buttonReleased];
                    break;
                case iCadeButtonC:
                    [[weakSelf.iCadeGamepad buttonB] buttonReleased];
                    break;
                case iCadeButtonD:
                    [[weakSelf.iCadeGamepad buttonY] buttonReleased];
                    break;
                case iCadeButtonE:
                    [[weakSelf.iCadeGamepad leftShoulder] buttonReleased];
                    break;
                case iCadeButtonF:
                    [[weakSelf.iCadeGamepad rightShoulder] buttonReleased];
                    break;
                case iCadeButtonG:
                    [[weakSelf.iCadeGamepad leftTrigger] buttonReleased];
                    break;
                case iCadeButtonH:
                    [[weakSelf.iCadeGamepad rightTrigger] buttonReleased];
                    break;
                case iCadeJoystickDown:
                case iCadeJoystickLeft:
                case iCadeJoystickRight:
                case iCadeJoystickUp:
                    [[weakSelf.iCadeGamepad dpad] padChanged];
                    break;
                default:
                    break;
            }
        };
    }
    return self;
}

- (NSString *) vendorName {
    return @"Mocute";
}

@end

File: Provenance/Controller/iCade/PViCadeMocuteController.swift
//
//  PViCadeMocuteController.swift
//  Provenance
//
//  Created by Joseph Mattiello on 4/12/18.
//  Copyright © 2018 Provenance. All rights reserved.
//

public class PViCadeMocuteController : PViCadeController {

	override init() {
        super.init()

		reader.buttonDown = {[weak self] (_ button: iCadeState) -> Void in
            switch button {
                case iCadeButtonA:
                    self?.iCadeGamepad.buttonX.buttonPressed()
                case iCadeButtonB:
                    self?.iCadeGamepad.buttonA.buttonPressed()
                case iCadeButtonC:
                    self?.iCadeGamepad.buttonB.buttonPressed()
                case iCadeButtonD:
                    self?.iCadeGamepad.buttonY.buttonPressed()
                case iCadeButtonE:
                    self?.iCadeGamepad.rightShoulder.buttonPressed()
                case iCadeButtonF:
                    self?.iCadeGamepad.leftShoulder.buttonPressed()
                case iCadeButtonG:
                    self?.iCadeGamepad.leftTrigger.buttonPressed()
                case iCadeButtonH:
                    self?.iCadeGamepad.rightTrigger.buttonPressed()
                case iCadeJoystickDown, iCadeJoystickLeft, iCadeJoystickRight, iCadeJoystickUp:
                    self?.iCadeGamepad.dpad().padChanged()
                default:
                    break
            }
            if self?.controllerPressedAnyKey != nil {
                self?.controllerPressedAnyKey(self)
            }
        }

        reader.buttonUp = {[weak self] (_ button: iCadeState) -> Void in
            switch button {
                case iCadeButtonA:
                    self?.iCadeGamepad.buttonX.buttonReleased()
                case iCadeButtonB:
                    self?.iCadeGamepad.buttonA.buttonReleased()
                case iCadeButtonC:
                    self?.iCadeGamepad.buttonB.buttonReleased()
                case iCadeButtonD:
                    self?.iCadeGamepad.buttonY.buttonReleased()
                case iCadeButtonE:
                    self?.iCadeGamepad.rightShoulder.buttonReleased()
                case iCadeButtonF:
                    self?.iCadeGamepad.leftShoulder.buttonReleased()
                case iCadeButtonG:
                    self?.iCadeGamepad.leftTrigger.buttonReleased()
                case iCadeButtonH:
                    self?.iCadeGamepad.rightTrigger.buttonReleased()
                case iCadeJoystickDown, iCadeJoystickLeft, iCadeJoystickRight, iCadeJoystickUp:
                    self?.iCadeGamepad.dpad().padChanged()
                default:
                    break
            }
        }
    
    }
    func vendorName() -> String? {
        return "Mocute"
    }
}
